src/lib/rnp2.c

```diff
@@ -1223,11 +1288,451 @@ rnp_op_encrypt_destroy(rnp_op_encrypt_t op)
+    (*op)->ffi = ffi;
+    (*op)->input = input;
+    (*op)->output = output;
+
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_verify_detached_create(rnp_op_verify_t *op,
+                              rnp_ffi_t        ffi,
+                              rnp_input_t      input,
+                              rnp_input_t      signature)
+{
+    if (!op || !ffi || !input || !signature) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+
+    *op = calloc(1, sizeof(**op));
+    if (!*op) {
+        return RNP_ERROR_OUT_OF_MEMORY;
+    }
+
+    rnp_ctx_init_ffi(&(*op)->rnpctx, ffi);
+    (*op)->ffi = ffi;
+    (*op)->input = signature;
+    (*op)->detached_input = input;
+
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_verify_execute(rnp_op_verify_t op)
+{
+    pgp_parse_handler_t handler = {
+      .key_provider =
+        &(pgp_key_provider_t){.callback = key_provider_bounce, .userdata = op->ffi},
+      .on_signatures = rnp_op_verify_on_signatures,
+      .src_provider = rnp_verify_src_provider,
+      .dest_provider = rnp_verify_dest_provider,
+      .param = op,
+      .ctx = &op->rnpctx};
+
+    rnp_result_t ret = process_pgp_source(&handler, &op->input->src);
+
+    if (op->output) {
+        op->output->keep = ret == RNP_SUCCESS;
+    }
+
+    return ret;
+}
+
+rnp_result_t
+rnp_op_verify_get_signature_count(rnp_op_verify_t op, size_t *count)
+{
+    if (!op || !count) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+
+    *count = op->signature_count;
+
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_verify_get_signature_at(rnp_op_verify_t op, size_t idx, rnp_op_verify_signature_t *sig)
+{
+    if (!op || !sig) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+
+    if (idx >= op->signature_count) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+
+    *sig = &op->signatures[idx];
+
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_verify_get_file_info(rnp_op_verify_t op,
+                            char *          filename_buf,
+                            size_t *        filename_buf_sz,
+                            uint32_t *      file_mtime)
+{
+    if (file_mtime) {
+        *file_mtime = op->file_mtime;
+    }
+
+    if (filename_buf_sz) {
+        size_t filename_len = strlen(op->filename);
+        size_t avail = *filename_buf_sz;
+
+        *filename_buf_sz = filename_len;
+
+        if (filename_buf && filename_len < avail) {
+            strcpy(filename_buf, op->filename);
+        }
+    }
+
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_verify_destroy(rnp_op_verify_t op)
+{
+    if (op) {
+        rnp_ctx_free(&op->rnpctx);
+        free(op->signatures);
+        free(op);
+    }
+
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_verify_signature_get_status(rnp_op_verify_signature_t sig)
+{
+    if (!sig) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+
+    return sig->verify_status;
+}
+
+rnp_result_t
+rnp_op_verify_signature_get_hash_fn(rnp_op_verify_signature_t sig,
```

Comment:
This is a bit inconsistent on memory handling. Most of the other functions will return allocated memory which must be freed with `rnp_buffer_free`

We have also talked about avoiding in/out params before. I know botan uses this convention but @ronaldtse has [mentioned](https://github.com/riboseinc/rnp/pull/319#issuecomment-316705575) that it might be good to avoid this (I think I dropped the ball on adding it to the dev guide though).

Reply:
Ok, agree.